src/engine.rs

```diff
@@ -190,144 +182,176 @@ where
     ) {
-        match content {
+        let memtable = memtables.get_or_insert(item.raft_group_id);
+        match item.content {
             LogItemContent::Entries(entries_to_add) => {
-                memtable.append(
-                    entries_to_add.entries,
-                    entries_to_add.entries_index.into_inner(),
-                );
+                let entries = entries_to_add.entries;
+                let mut entries_index = entries_to_add.entries_index.into_inner();
+                if queue == LogQueue::Rewrite {
+                    // By default the queue is for appending.
+                    entries_index.iter_mut().for_each(|ei| ei.queue = queue);
+                }
+                memtable.wl().append(entries, entries_index);
             }
             LogItemContent::Command(Command::Clean) => {
-                memtable.remove();
+                memtables.remove(item.raft_group_id);
             }
             LogItemContent::Command(Command::Compact { index }) => {
-                memtable.compact_to(index);
+                memtable.wl().compact_to(index);
             }
             LogItemContent::Kv(kv) => match kv.op_type {
-                OpType::Put => memtable.put(kv.key, kv.value.unwrap(), file_num),
-                OpType::Del => memtable.delete(kv.key.as_slice()),
+                OpType::Put => {
+                    let (key, value) = (kv.key, kv.value.unwrap());
+                    match queue {
+                        LogQueue::Append => memtable.wl().put(key, value, file_num),
+                        LogQueue::Rewrite => memtable.wl().put_rewrite(key, value, file_num),
+                    }
+                }
+                OpType::Del => memtable.wl().delete(kv.key.as_slice()),
             },
         }
     }
 
+    fn rewrite_log_item_to_memtable(
+        memtables: &MemTableAccessor<E, W>,
+        item: LogItem<E>,
+        file_num: u64,
+        latest_rewrite: u64,
+    ) {
+        let memtable = memtables.get_or_insert(item.raft_group_id);
+        match item.content {
+            LogItemContent::Entries(entries_to_add) => {
+                let entries_index = entries_to_add.entries_index.into_inner();
+                memtable.wl().rewrite(entries_index, latest_rewrite);
+            }
+            LogItemContent::Kv(kv) => match kv.op_type {
+                OpType::Put => memtable.wl().rewrite_key(kv.key, latest_rewrite, file_num),
+                _ => unreachable!(),
+            },
+            _ => unreachable!(),
+        }
+    }
+
     fn apply_to_memtable(
         memtables: &MemTableAccessor<E, W>,
         log_batch: LogBatch<E, W>,
+        queue: LogQueue,
         file_num: u64,
     ) {
         for item in log_batch.items {
-            let m = memtables.get_or_insert(item.raft_group_id);
-            Self::apply_log_item_to_memtable(item.content, &mut *m.wl(), file_num);
+            Self::apply_log_item_to_memtable(memtables, item, queue, file_num);
```

Comment:
Could we move the method `apply_log_item_to_memtable` as a member method of `MemTableAccessor `?

Reply:
I make it as a normal function.